### PR TITLE
Avoid bucket copies in Aggregations

### PR DESCRIPTION
**Pull request by original-brownbear**
Motivated by heap dumps for aggregations often containing mostly duplicate bucket instances.
Note that This is a start, but there's a lot of duplication we can remove in follow-ups.
But for this one, there's simply a lot of straight forward spots where we can avoid copying a bucket/list-of-buckets and derived instances.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java [hunk 1]
```diff
@@ -236,11 +236,34 @@ public static InternalAggregations topLevelReduce(List<InternalAggregations> agg
...
+            for (int i = 0; i < reducedInternalAggs.size(); i++) {
+                InternalAggregation agg = reducedInternalAggs.get(i);
+                InternalAggregation internalAggregation = agg.reducePipelines(
+                    agg,
+                    context,
+                    context.pipelineTreeRoot().subTree(agg.getName())
+                );
+                if (internalAggregation.equals(agg) == false) {
+                    if (internalAggregations == null) {
+                        internalAggregations = new ArrayList<>(reducedInternalAggs);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java:249]: No need to be tricky here IMO, copying the full list is likely similar in cost to doing anything tricky like copying the first I unchanged elements and then moving to a different loop or so, given how short these lists are.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AbstractHistogramAggregator.java [hunk 2]
```diff
@@ -84,6 +84,9 @@ public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws I
             double key = roundKey * interval + offset;
             return new InternalHistogram.Bucket(key, docCount, keyed, formatter, subAggregationResults);
         }, (owningBucketOrd, buckets) -> {
+            if (buckets.isEmpty()) {
+                return buildEmptyAggregation();
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AbstractHistogramAggregator.java:88]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java [hunk 3]
```diff
@@ -340,6 +340,9 @@ public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws I
         return buildAggregationsForVariableBuckets(owningBucketOrds, bucketOrds, (bucketValue, docCount, subAggregationResults) -> {
             return new InternalDateHistogram.Bucket(bucketValue, docCount, keyed, formatter, subAggregationResults);
         }, (owningBucketOrd, buckets) -> {
+            if (buckets.isEmpty()) {
+                return buildEmptyAggregation();
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java:344]: :+1:

**Review by iverase — APPROVED** [review]
lgtm

**Comment by original-brownbear** [comment]
Thanks Ignacio!

**Comment by benwtrent** [comment]
@original-brownbear @iverase do we know if this has any measurable improvement in memory usage or CPU?

This change is causing some nasty bugs and we need to make sure either we back this change out, or fix the individual bugs.

**Comment by original-brownbear** [comment]
@benwtrent 

> do we know if this has any measurable improvement in memory usage or CPU?

Yes it should, if I remember correctly this was motivated by a heap dump that showed O(100M) potential savings here. I'll try to look into this and see if we can make this safer easily, sorry for the bugs :/ Reverting would be a reasonable plan B here though IMO, there's so many issues in eggs and this I just one of them :)

**Comment by nik9000** [comment]
I'm going to revert this change now and we can tie it to a rework on `equals` for aggs to make it harder to make mistakes with them. I'm not sure we want to make that change at all, but this change is going to be locked behind it.

**Comment by benwtrent** [comment]
@original-brownbear the only way to make this change work is to enforce that `equals` is overridden, this will cause plugins to fail to compile, so we need to revert this change, enforce `equals`, and then we can add it back.

^ Yeah, what @nik9000 said.

**Pull request merged**